File: it.unibo.enablerCleanArch/src/it/unibo/enablerCleanArch/domain/ApplMessage.java
package it.unibo.enablerCleanArch.domain;

import alice.tuprolog.Struct;
import alice.tuprolog.Term;

enum ApplMessageType{
    event, dispatch, request, reply, invitation
}

public class ApplMessage {
    protected String msgId       = "";
    protected String msgType     = null;
    protected String msgSender   = "";
    protected String msgReceiver = "";
    protected String msgContent  = "";
    protected int msgNum         = 0;
	
	public ApplMessage( 
			String MSGID, String MSGTYPE, String SENDER, String RECEIVER, String CONTENT, String SEQNUM ) {
        msgId 		= MSGID;
        msgType 	= MSGTYPE;
        msgSender 	= SENDER;
        msgReceiver = RECEIVER;
        msgContent 	= CONTENT;
        msgNum = Integer.parseInt(SEQNUM);		
	}
	
	
	
    public ApplMessage( String msg ) {
        //msg( MSGID, MSGTYPE, SENDER, RECEIVER, CONTENT, SEQNUM )
    	Struct msgStruct = (Struct) Term.createTerm(msg);
        setFields(msgStruct);
    }
	
    private void setFields( Struct msgStruct ) {
        msgId 		= msgStruct.getArg(0).toString();
        msgType 	= msgStruct.getArg(1).toString();
        msgSender 	= msgStruct.getArg(2).toString();
        msgReceiver = msgStruct.getArg(3).toString();
        msgContent 	= msgStruct.getArg(4).toString();
        msgNum 		= Integer.parseInt(msgStruct.getArg(5).toString());
    }

    public String msgId() {   return msgId; }
    public String msgType() { return msgType; }
    public String msgSender() { return msgSender; }
    public String msgReceiver() { return msgReceiver;  }
    public String msgContent() { return msgContent;  }
    public String msgNum() { return "" + msgNum; }
    
    
    public boolean isEvent(){
        return msgType == ApplMessageType.event.toString();
    }
    public boolean isDispatch(){
        return msgType == ApplMessageType.dispatch.toString();
    }
    public boolean isRequest(){
        return msgType == ApplMessageType.request.toString();
    }
    public boolean isInvitation(){
        return msgType == ApplMessageType.invitation.toString();
    }
    public boolean isReply(){
        return msgType == ApplMessageType.reply.toString();
    }   

}

File: it.unibo.kotlinSupports/app/src/main/kotlin/it/unibo/actor0/ApplMessage.kt
package it.unibo.actor0

import it.unibo.`is`.interfaces.protocols.IConnInteraction
import it.unibo.robotService.ApplMsgs

enum class ApplMessageType{
    event, dispatch, request, reply, invitation
}
//msg(guiCmd,dispatch,userConsole,any,stop,1)
open class ApplMessage(
        val msgId: String = "", val msgType: String = ApplMessageType.dispatch.toString(),
        val msgSender: String = "", val msgReceiver: String = "",
        val msgContent: String = "", val msgNum: String = "0"
)  {

    companion object{
        @JvmStatic
        //msg( MSGID, MSGTYPE, SENDER, RECEIVER, CONTENT, SEQNUM )
        fun create(msg: String) : ApplMessage{
            //sysUtil.colorPrint( "ApplMessage | CREATE: $msg " )
            val jsonContent = msg.split("{")[1].split("}")[0]
            val content = "{$jsonContent}"
            //sysUtil.colorPrint( "ApplMessage | jsonContent: $jsonContent " )
            val msgNoContent = msg.replace(jsonContent,"xxx")  //xxx will become item[4]
            //sysUtil.colorPrint( "ApplMessage | msgNoContent=$msgNoContent " )
            val body  = msgNoContent.replace("msg","").replace("(","").replace(")","")
            val items = body.split(",")
            return ApplMessage(items[0],items[1], items[2], items[3], content, items[5] )
        }
    }

    var conn : IConnInteraction? = null

    fun isEvent(): Boolean{
        return msgType == ApplMessageType.event.toString()
    }
    fun isDispatch(): Boolean{
        return msgType == ApplMessageType.dispatch.toString()
    }
    fun isRequest(): Boolean{
        return msgType == ApplMessageType.request.toString()
    }
    fun isReply(): Boolean{
        return msgType == ApplMessageType.reply.toString()
    }

    override fun toString(): String {
        return getDefaultRep()
    }
    fun getDefaultRep(): String {
        return "msg($msgId,$msgType,$msgSender,$msgReceiver,$msgContent,${msgNum})"
        //if (msgType == null) "msg(none,none,none,none,none,0)" else

    }



}//ApplMessage

//just to test
fun main(){

    //val m = ApplMessage.create( ApplMsgs.testCreateMsg().toString() )
    val m = "msg(robotmove,dispatch,actortry,stepRobot,{\"robotmove\":\"turnLeft\", \"time\": 300},1)"
    println(m)

}